channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
+    let default_lang = String::from("en");
+    let mut lang_tree: BTreeMap<String, String> = BTreeMap::new();
+    let mut i = 0;
+    alheader.split(",").for_each(|l| {
+        if l.contains(";") {
+            let weight: Vec<&str> = l.split(";").collect();
+            let lang = weight[0].to_ascii_lowercase();
+            let pref = weight[1].to_ascii_lowercase();
+            lang_tree.insert(String::from(pref), String::from(lang));
+        } else {
+            lang_tree.insert(
+                format!("q=1.{:02}", i),
+                String::from(l.to_ascii_lowercase()),
+            );
+            i += 1;
+        }
+    });
+    let mut langs: Vec<String> = lang_tree.values().map(|l| l.to_owned()).collect();
+    langs.reverse();
+    langs.push(default_lang);
+    langs
+}
+
+// Return the element that most closely matches the preferred language.
+// This rounds up from the dialect if possible.
+fn get_preferred_language_element(
+    langs: &Vec<String>,
+    elements: BTreeMap<String, String>,
+) -> Option<String> {
+    for lang in langs.clone() {
+        if elements.contains_key(&lang) {
+            //unwrap is safe-ish because we've checked that the key exists.
+            return Some(elements.get(lang.as_str()).unwrap().to_string());
+        }
+        if lang.contains("-") {
+            let (lang, _) = lang.split_at(2);
+            if elements.contains_key(lang) {
+                return Some(elements.get(lang).unwrap().to_string());
+            }
+        }
+    }
+    None
+}
+
+// Set the sender meta information from the request headers.
+impl From<HttpRequest<WsChannelSessionState>> for SenderData {
+    fn from(req: HttpRequest<WsChannelSessionState>) -> Self {
+        let mut sender = SenderData::default();
+        let headers = req.headers();
+        //TODO: Get the default lang
+        let langs = match headers.get(http::header::ACCEPT_LANGUAGE) {
+            None => vec![String::from("en")],
+            Some(l) => preferred_language(l.to_str().unwrap_or("").to_owned()),
+        };
+        let conn = req.connection_info();
+        // parse user-header for platform info
+        sender.ua = match headers.get(http::header::USER_AGENT) {
+            None => None,
+            Some(s) => match s.to_str() {
+                Err(x) => {
+                    println!("Error: bad UA string: {:?}", x);
+                    None
+                }
+                Ok(s) => Some(s.to_owned()),
+            },
+        };
+        sender.addr = match conn.remote() {
+            Some(a) => Some(a.to_owned()),
+            None => None,
+        };
+        if sender.addr.is_some() {
+            if let Ok(loc) = sender.addr.clone().unwrap().parse() {
+                if let Ok(city) = req.state().iploc.lookup::<City>(loc) {
```

Comment:
Would we ever expect `iploc.lookup` to return an "expected error" in case e.g. there's no location data for the given address?  Or, should any errors returned here be treated as an noteworthy exception and logged somehow for future reference?  I'd hate for a corrupted database to just silently fail to ever produce lookup info when it should probably spew very noisily to the logs.

Reply:
From previous experience, there can be a lot of IPs that return no geo information. It can really clog up the logs. I can look at trying to add a metric for this, though.

Reply:
Fair point. I may filter out the "Not Founds" and elevate the others with a match.